File: services/aitec-engine/app/telemetry.py
from __future__ import annotations

import asyncio
import json
import os
import re
import secrets
import threading
import time
import urllib.request
from collections import defaultdict

HTTP_BUCKETS=(0.005,0.01,0.025,0.05,0.1,0.25,0.5,1.0,1.5,2.5,5.0,10.0)
_lock=threading.Lock()
_active=0
_stats=defaultdict(lambda:{'count':0,'sum':0.0,'buckets':[0 for _ in HTTP_BUCKETS]})
_traceparent_re=re.compile(r'^00-([0-9a-fA-F]{32})-([0-9a-fA-F]{16})-[0-9a-fA-F]{2}$')


def _trace_ids(header:str|None):
    match=_traceparent_re.match(header or '')
    trace_id=(match.group(1).lower() if match else secrets.token_hex(16))
    return trace_id,secrets.token_hex(8)


def _record(method:str,status:int,duration:float):
    global _active
    status_class=f'{max(0,min(9,status//100))}xx'
    key=(method.upper()[:16],status_class)
    with _lock:
        _active=max(0,_active-1)
        item=_stats[key]
        item['count']+=1
        item['sum']+=max(0.0,duration)
        for index,limit in enumerate(HTTP_BUCKETS):
            if duration<=limit:item['buckets'][index]+=1


def _send_span(base:str,service:str,trace_id:str,span_id:str,start_ns:int,end_ns:int,method:str,path:str,status:int):
    payload={'resourceSpans':[{'resource':{'attributes':[{'key':'service.name','value':{'stringValue':service}},{'key':'service.version','value':{'stringValue':'19.0.0-rc.3'}}]},'scopeSpans':[{'scope':{'name':'lotediretor.native-http'},'spans':[{'traceId':trace_id,'spanId':span_id,'name':f'{method} {path}','kind':2,'startTimeUnixNano':str(start_ns),'endTimeUnixNano':str(end_ns),'attributes':[{'key':'http.request.method','value':{'stringValue':method}},{'key':'url.path','value':{'stringValue':path}},{'key':'http.response.status_code','value':{'intValue':str(status)}}],'status':{'code':2 if status>=500 else 1}}]}]}]}
    try:
        request=urllib.request.Request(base.rstrip('/')+'/v1/traces',data=json.dumps(payload).encode('utf-8'),headers={'content-type':'application/json'},method='POST')
        with urllib.request.urlopen(request,timeout=1.5) as response:
            response.read(1)
    except Exception:
        pass


async def http_telemetry(request,call_next):
    global _active
    method=str(request.method or 'HTTP').upper()
    path=str(request.url.path or '/')
    trace_id,span_id=_trace_ids(request.headers.get('traceparent'))
    start_perf=time.perf_counter()
    start_ns=time.time_ns()
    with _lock:_active+=1
    try:
        response=await call_next(request)
        status=int(response.status_code)
    except Exception:
        status=500
        duration=time.perf_counter()-start_perf
        _record(method,status,duration)
        base=os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT','').strip()
        if base:
            asyncio.create_task(asyncio.to_thread(_send_span,base,'aitec-engine',trace_id,span_id,start_ns,time.time_ns(),method,path,status))
        raise
    duration=time.perf_counter()-start_perf
    _record(method,status,duration)
    response.headers['traceparent']=f'00-{trace_id}-{span_id}-01'
    response.headers['x-trace-id']=trace_id
    base=os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT','').strip()
    if base:
        asyncio.create_task(asyncio.to_thread(_send_span,base,'aitec-engine',trace_id,span_id,start_ns,time.time_ns(),method,path,status))
    return response


def render_metrics(service:str):
    with _lock:
        active=_active
        snapshot={key:{'count':value['count'],'sum':value['sum'],'buckets':list(value['buckets'])} for key,value in _stats.items()}
    lines=[
        '# HELP lotediretor_http_requests_total Completed HTTP requests.',
        '# TYPE lotediretor_http_requests_total counter',
        '# HELP lotediretor_http_request_duration_seconds HTTP request duration histogram.',
        '# TYPE lotediretor_http_request_duration_seconds histogram',
        '# HELP lotediretor_http_requests_active HTTP requests currently in flight.',
        '# TYPE lotediretor_http_requests_active gauge',
        f'lotediretor_http_requests_active{{service="{service}"}} {active}',
    ]
    for (method,status_class),item in sorted(snapshot.items()):
        labels=f'service="{service}",method="{method}",status_class="{status_class}"'
        lines.append(f'lotediretor_http_requests_total{{{labels}}} {item["count"]}')
        for limit,count in zip(HTTP_BUCKETS,item['buckets']):
            lines.append(f'lotediretor_http_request_duration_seconds_bucket{{{labels},le="{limit}"}} {count}')
        lines.append(f'lotediretor_http_request_duration_seconds_bucket{{{labels},le="+Inf"}} {item["count"]}')
        lines.append(f'lotediretor_http_request_duration_seconds_sum{{{labels}}} {item["sum"]:.9f}')
        lines.append(f'lotediretor_http_request_duration_seconds_count{{{labels}}} {item["count"]}')
    return '\n'.join(lines)

File: services/control-api/src/telemetry.ts
import {CONTROL_VERSION} from './version';
import {randomBytes} from 'crypto';

type TraceState={traceId:string;spanId:string;startNs:bigint;name:string};
type HttpMetric={count:number;sumSeconds:number;buckets:number[]};

const HTTP_BUCKETS=[0.005,0.01,0.025,0.05,0.1,0.25,0.5,1,1.5,2.5,5,10];
const HTTP_METRICS=new Map<string,HttpMetric>();
let ACTIVE_REQUESTS=0;

function hex(n:number){return randomBytes(n).toString('hex')}
function metricKey(method:string,statusClass:string){return `${method}|${statusClass}`}
function escapeLabel(value:string){return value.replace(/\\/g,'\\\\').replace(/"/g,'\\"').replace(/\n/g,'\\n')}

function recordHttp(methodRaw:any,status:number,durationSeconds:number){
  const method=String(methodRaw||'UNKNOWN').toUpperCase().slice(0,16);
  const statusClass=Number.isFinite(status)?`${Math.max(0,Math.min(9,Math.floor(status/100)))}xx`:'unknown';
  const key=metricKey(method,statusClass);
  const metric=HTTP_METRICS.get(key)||{count:0,sumSeconds:0,buckets:HTTP_BUCKETS.map(()=>0)};
  metric.count+=1;
  metric.sumSeconds+=Math.max(0,durationSeconds);
  HTTP_BUCKETS.forEach((le,index)=>{if(durationSeconds<=le)metric.buckets[index]+=1});
  HTTP_METRICS.set(key,metric);
}

export function beginHttpTrace(req:any):TraceState{
  const tp=String(req.headers?.traceparent||'');const m=/^00-([0-9a-f]{32})-([0-9a-f]{16})-[0-9a-f]{2}$/i.exec(tp);
  const state={traceId:m?.[1]?.toLowerCase()||hex(16),spanId:hex(8),startNs:process.hrtime.bigint(),name:`${req.method||'HTTP'} ${req.url||'/'}`};
  ACTIVE_REQUESTS+=1;
  req.ldTrace=state;req.traceId=state.traceId;return state;
}
export function traceparent(s:TraceState){return `00-${s.traceId}-${s.spanId}-01`}

export function finishHttpTrace(service:string,req:any,reply:any){
  const s=req.ldTrace as TraceState|undefined;
  if(!s)return;
  const elapsedNs=process.hrtime.bigint()-s.startNs;
  const durationSeconds=Number(elapsedNs)/1e9;
  const status=Number(reply.statusCode||200);
  ACTIVE_REQUESTS=Math.max(0,ACTIVE_REQUESTS-1);
  recordHttp(req.method,status,durationSeconds);

  const base=(process.env.OTEL_EXPORTER_OTLP_ENDPOINT||'').replace(/\/$/,'');
  if(!base)return;
  const endUnixNs=BigInt(Date.now())*1000000n;
  const startUnixNs=endUnixNs-elapsedNs;
  const payload={resourceSpans:[{resource:{attributes:[{key:'service.name',value:{stringValue:service}},{key:'service.version',value:{stringValue:CONTROL_VERSION}}]},scopeSpans:[{scope:{name:'lotediretor.native-http'},spans:[{traceId:s.traceId,spanId:s.spanId,name:s.name,kind:2,startTimeUnixNano:String(startUnixNs),endTimeUnixNano:String(endUnixNs),attributes:[{key:'http.request.method',value:{stringValue:String(req.method||'')}},{key:'url.path',value:{stringValue:String(req.url||'').split('?')[0]}},{key:'http.response.status_code',value:{intValue:String(status)}}],status:{code:status>=500?2:1}}]}]}]};
  const ac=new AbortController();const timer=setTimeout(()=>ac.abort(),1500);
  void fetch(`${base}/v1/traces`,{method:'POST',headers:{'content-type':'application/json'},body:JSON.stringify(payload),signal:ac.signal}).catch(()=>{}).finally(()=>clearTimeout(timer));
}

export function renderHttpMetrics(service:string){
  const s=escapeLabel(service);
  const lines=[
    '# HELP lotediretor_http_requests_total Completed HTTP requests.',
    '# TYPE lotediretor_http_requests_total counter',
    '# HELP lotediretor_http_request_duration_seconds HTTP request duration histogram.',
    '# TYPE lotediretor_http_request_duration_seconds histogram',
    '# HELP lotediretor_http_requests_active HTTP requests currently in flight.',
    '# TYPE lotediretor_http_requests_active gauge',
    `lotediretor_http_requests_active{service="${s}"} ${ACTIVE_REQUESTS}`,
  ];
  for(const [key,metric] of [...HTTP_METRICS.entries()].sort(([a],[b])=>a.localeCompare(b))){
    const [method,statusClass]=key.split('|').map(escapeLabel);
    const labels=`service="${s}",method="${method}",status_class="${statusClass}"`;
    lines.push(`lotediretor_http_requests_total{${labels}} ${metric.count}`);
    HTTP_BUCKETS.forEach((le,index)=>lines.push(`lotediretor_http_request_duration_seconds_bucket{${labels},le="${le}"} ${metric.buckets[index]}`));
    lines.push(`lotediretor_http_request_duration_seconds_bucket{${labels},le="+Inf"} ${metric.count}`);
    lines.push(`lotediretor_http_request_duration_seconds_sum{${labels}} ${metric.sumSeconds.toFixed(9)}`);
    lines.push(`lotediretor_http_request_duration_seconds_count{${labels}} ${metric.count}`);
  }
  return lines.join('\n');
}
